nestkernel/source_table.cpp

```diff
@@ -320,67 +364,59 @@ nest::SourceTable::get_next_target_data( const thread tid,
     current_source.set_processed( true );
 
     // we need to set a marker stating whether the entry following this
-    // entry, if existent, has the same source; start by assuming it
-    // has a different source, only change if necessary
-    kernel().connection_manager.set_has_source_subsequent_targets(
-      current_position.tid, current_position.syn_id, current_position.lcid, false );
-    if ( ( current_position.lcid + 1
-             < static_cast< long >( sources_[ current_position.tid ][ current_position.syn_id ].size() )
-           and sources_[ current_position.tid ][ current_position.syn_id ][ current_position.lcid + 1 ].get_node_id()
-             == current_source.get_node_id() ) )
+    // entry, if existent, has the same source
+    kernel().connection_manager.set_has_source_subsequent_targets( current_position.tid,
+      current_position.syn_id,
+      current_position.lcid,
+      next_entry_has_same_source( current_position, current_source ) );
+
+    // no need to communicate this entry if the previous entry has the
+    // same source
+    if ( previous_entry_has_same_source( current_position, current_source ) )
     {
-      kernel().connection_manager.set_has_source_subsequent_targets(
-        current_position.tid, current_position.syn_id, current_position.lcid, true );
+      current_position.decrease();
+      continue;
     }
 
-    // decrease the position without returning a TargetData if the
-    // entry preceding this entry has the same source, but only if
-    // the preceding entry was not processed yet
-    if ( ( current_position.lcid - 1 >= 0 )
-      and ( sources_[ current_position.tid ][ current_position.syn_id ][ current_position.lcid - 1 ].get_node_id()
-            == current_source.get_node_id() )
-      and ( not sources_[ current_position.tid ][ current_position.syn_id ][ current_position.lcid - 1 ]
-                  .is_processed() ) )
+    // reaching this means we found a valid TargetData
+
+    // set the source rank
+    source_rank = kernel().mpi_manager.get_process_id_of_node_id( current_source.get_node_id() );
```

Comment:
I'd introduce
```c++
const auto cs_gid =current_source.get_node_id();
```
to simplify the lines below.

Reply:
ok